staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -330,7 +330,15 @@ func (s *store) conditionalDelete(
 			origStateIsCurrent = true
 			continue
 		}
-		return decode(s.codec, s.versioner, origState.data, out, origState.rev)
+
+		if len(txnResp.Responses) == 0 || txnResp.Responses[0].GetResponseDeleteRange() == nil {
+			return errors.New(fmt.Sprintf("invalid DeleteRange response: %v", txnResp.Responses))
```

Comment:
Making these errors turns what would previously have been a successful delete response into an error. Is it better to error so callers don't get incorrect resourceVersions, or return origState.rev in this case so callers don't get errors?

Reply:
I was thinking about that, and my personal opinion is to rather return an error in this case.
I think that neither of these should generally happen looking at the implementation (though as you pointed out some corner cases are in theory possible).

But in those cases, I think it's better to error, than to send RV that would be misleading to users expect the semantic of "this is the current RV".

If you think very strongly the other way I can change that, but I think the current way is better (note that this code path is not used in watch, so it won't result in any broken watches or additional relists or anything like that - it's only the explicit Delete code path).

Reply:
What would it even mean to get a successful transaction against etcd, with no response and no recorded delete? Would we even be certain that the deletion occurred? Would we try to undo it?

Reply:
I don't know what I don't know about possible etcd responses... is it documented what we can rely on from successful delete transactions? If non-nil/non-zero `.Responses[0].GetResponseDeleteRange().Header.Revision` is guaranteed, erroring here seems fine

Reply:
ok, I'm just scarred from things like https://github.com/kubernetes/kubernetes/pull/76675/files#diff-c84f7808522d6cd14f52c2d800c897bbff103fcee2c73fa3bbb13da81db72298 where we thought aspects of the etcd response were guaranteed and it turns out they were not in rare scenarios

if the header/rev in the response is reliable, this lgtm

Reply:
Sure - I understand that. The PrevKV is somewhat special, because you have to explicitly request it.

Here, all we need should be set (assuming no bugs in etcd, so having error handling is definitely reasonable to have).

Reply:
For posterity, about explicitly requesting PrevKV:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L236